Fix #6887: Audit jobs and tests for user models and a new panel for prod jobs on admin tab

ankita240796 opened the pull request:
##   Explanation
- Adds audit jobs and tests for user models
- Fixes #6887: Adds a separate panel for prod jobs on admin tab

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

ankita240796 committed:
Domain object changes for topic, skill, ques models

ankita240796 committed:
Fix test

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into domain-fix

ankita240796 committed:
Ensure strings convertible to ckeditor dont fail validation

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into domain-fix

ankita240796 committed:
Fix lint

ankita240796 committed:
Resolve conflicts

ankita240796 committed:
Address review comments

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into domain-fix

ankita240796 committed:
Change id to misconception id

ankita240796 committed:
Add id check

ankita240796 committed:
Remove extra attr

ankita240796 committed:
Fix test

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into domain-fix

ankita240796 committed:
Fix issues

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into domain-fix

ankita240796 committed:
Fix test

ankita240796 committed:
Add test for all commit log models

ankita240796 committed:
Update related model to external model

ankita240796 committed:
Fix check

ankita240796 committed:
User models

ankita240796 committed:
Resolve conflicts

ankita240796 committed:
Resolve conflicts

ankita240796 committed:
Fix test

seanlip commented on core/domain/prod_validation_jobs_one_off.py:
```diff
@@ -529,6 +544,117 @@ def validate(cls, item):
+
+    @classmethod
+    def _get_model_id_regex(cls, unused_item):
+        return '^\\d*$'
```
Can you explain the justification for this regex? Why * and not +?

ankita240796 replied:
My bad, it will be `+`. Fixed this.

seanlip commented on core/domain/prod_validation_jobs_one_off.py:
```diff
@@ -529,6 +544,117 @@ def validate(cls, item):
+            cls, unused_item):
+        """Returns a list common properties to dismatch.
+
+        # This can be overridden by subclasses, if needed.
```
Drop `# ` at start

ankita240796 replied:
Done.

seanlip commented on core/domain/prod_validation_jobs_one_off.py:
```diff
@@ -529,6 +544,117 @@ def validate(cls, item):
+    @classmethod
+    def _get_common_properties_of_external_model_which_should_not_match(
+            cls, unused_item):
+        """Returns a list common properties to dismatch.
```
I don't really understand this docstring. Can you elaborate on what "dismatch" / "should not match" means? What are the things that shouldn't match, exactly?

ankita240796 replied:
Updated, added an example to explain.

If two models store opposite features, the common properties in both the models should not have the same values. For example, in `CompletedActivitiesModel` and `IncompleteActivitiesModel` the `exploration_ids` in one model should not match those in the other model.

ankita240796 replied:
I am using `get_by_id` at two other places:

1. https://github.com/oppia/oppia/pull/7023/files#diff-ecafbbaeb117bcae187e78e5b33e120fR3989

This to obtain exploration models corresponding to node ids in StoryProgressModel. For this, I wil need to fetch story models and the get the exploration ids for completed nodes in StoryProgressModel. So, I don't think if we can separate this.

2. https://github.com/oppia/oppia/pull/7023/files#diff-ecafbbaeb117bcae187e78e5b33e120fR4080

This to obtain UserBulkEmailModels corresponding to recipients to whom bulk email is sent after a they qualify a query. Here also, I need to fetch the BulkEmailModel for each entity and iterate over recipients to fetch user bulk email models. So, the case becomes same as above.

What should I do in these two cases?

seanlip replied:
I am not entirely sure how your comment relates to my comment, but one observation I have is that you might want to use get_multi() in both cases rather than lots of get_by_id()'s in a for loop. The latter will take much longer.

ankita240796 replied:
Oh, sorry I added this in the wrong thread. I wanted to comment this [here](https://github.com/oppia/oppia/pull/7023#discussion_r297532713). I will change `get_by_id` to `get_multi`.

seanlip replied:
Just a note -- it's the repeated query() that we should try to avoid if possible (especially doing N queries when 1 will do).

get_by_id() is fast, but if you're doing multiple of them, do a single get_multi() instead.

The above principles should be sufficient for most cases.

ankita240796 replied:
Updated to `get_multi`.

seanlip commented on core/domain/prod_validation_jobs_one_off.py:
```diff
@@ -2474,213 +2600,1326 @@ def _get_custom_validation_functions(cls):
+        Args:
+            item: ndb.Model. UserSettingsModel to validate.
+        """
+        user_models_list = user_models.UserSettingsModel.query().filter(
```
This seems like a bad idea -- you're doing one query per user model, when you could instead just do one projection query for all the models, extract all the normalized usernames, sort them, and figure out if there are any duplicates.

ankita240796 replied:
So, should this be done in a separate job instead since it will not be dependent on a single model instance now?

seanlip replied:
Probably yes. I'm not sure if this is the case for others but it might be worth checking to see if you have similar patterns elsewhere. Doing N queries is quite resource-intensive.

ankita240796 replied:
I was trying to get all user models in a global list and then use it as follows:

```
ALL_USER_MODELS = user_models.UserSettingsModel.get_all().fetch()
USER_IDS_AND_NORMALIZED_USERNAMES = [
        (user_model.id, user_model.normalized_username)
        for user_model in ALL_USER_MODELS if user_model is not None]
```

But when I run the test, I get the error: `WARNING:root:initial generator _run_to_list(query.py:987) raised BadArgumentError(app must not be empty.)`. Is there any other way to fetch all the user models once for the job? 

I also found a solution for testing: https://www.dangtrinh.com/2013/12/google-app-engine-sdk-for-python-using.html but I am not sure if we should do this.

seanlip replied:
The solution mentioned in the last paragraph seems hacky, let's not do it.

For the first part -- I don't understand where you're adding the code. Is it at module-level? If so, that doesn't seem so good, it will get run when the module is imported which is probably not what you want.

Perhaps your best bet is to just have a special MR job that iterates through all the models and that yields the normalized usernames. Then have reduce() check that there are no duplicates.

ankita240796 replied:
Yes, I was adding the code at module level but I have changed it now to the solution you mentioned. Thanks for the solution!

seanlip commented on core/domain/prod_validation_jobs_one_off.py:
```diff
@@ -2474,213 +2600,1326 @@ def _get_custom_validation_functions(cls):
+                        key, feconf.DASHBOARD_STATS_DATETIME_STRING_FORMAT)
+                except Exception:
+                    is_invalid = True
+                if not is_invalid and key > current_time_str:
```
I think, instead of carrying is_invalid around, you could just use assertions within the try/except. That might lead to cleaner code.

ankita240796 replied:
Done.

seanlip commented on core/domain/prod_validation_jobs_one_off.py:
```diff
@@ -2474,213 +2600,1326 @@ def _get_custom_validation_functions(cls):
+        Args:
+            item: ndb.Model. ExplorationUserDataModel to validate.
+        """
+        if item.rating and (item.rating < 1 or item.rating > 5):
```
Are you allowing an item.rating of 0? If not, use `if item.rating is not None` rather than `if item.rating`.

Similarly elsewhere.

ankita240796 replied:
Fixed.

seanlip commented on core/domain/prod_validation_jobs_one_off.py:
```diff
@@ -2474,213 +2600,1326 @@ def _get_custom_validation_functions(cls):
+            error_msg = ''
+            if private_exp_ids:
+                error_msg = error_msg + (
+                    'Following exploration ids are private %s' % (
```
Put a `. ` after each error message, to separate them.

ankita240796 replied:
Done.

ankita240796 committed:
Address review comments

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into usermodels

ankita240796 commented:
Hi @seanlip, I have addressed the review comments, PTAL. Thanks!

ankita240796 committed:
Fix conflicts

ankita240796 committed:
Fix test

ankita240796 committed:
Resolve conflicts

seanlip commented:
Hi @ankita240796 -- just to check, I think [this comment](https://github.com/oppia/oppia/pull/7023#discussion_r297557814) and get_multi are still unaddressed right?

ankita240796 commented:
Hi @seanlip, Yes I had addressed the comments with my queries earlier. I will make the changes now and the ask for a final review again. Thanks!

Also, I have a doubt regarding the query for fetching all user models:

I was trying to get all user models in a global list and then use it as follows:

```
ALL_USER_MODELS = user_models.UserSettingsModel.get_all().fetch()
USER_IDS_AND_NORMALIZED_USERNAMES = [
        (user_model.id, user_model.normalized_username)
        for user_model in ALL_USER_MODELS if user_model is not None]
```

But when I run the test, I get the error: `WARNING:root:initial generator _run_to_list(query.py:987) raised BadArgumentError(app must not be empty.)`. Is there any other way to fetch all the user models once for the job? 

I also found a solution for testing: https://www.dangtrinh.com/2013/12/google-app-engine-sdk-for-python-using.html but I am not sure if we should do this.

ankita240796 committed:
Change get to get_multi

ankita240796 committed:
Make coverage 100%

ankita240796 committed:
Separate prod jobs on admin panel

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into usermodels

ankita240796 committed:
Add normalized username validation

ankita240796 commented:
Hi @seanlip, I have updated the code to remove get_by_id and multiple unrequired queries, PTAL. Thanks!

ankita240796 committed:
Make coverage 100%

ankita240796 committed:
Further increase coverage

seanlip commented on core/controllers/admin.py:
```diff
@@ -57,7 +57,9 @@ def get(self):
                 klass.__name__ == job['job_type']
-                for klass in jobs_registry.ONE_OFF_JOB_MANAGERS])
+                for klass in (
+                    jobs_registry.ONE_OFF_JOB_MANAGERS + (
```
The naming is a bit weird, because it seems like ONE_OFF_JOB_MANAGERS would include PROD_VALIDATION_ONE_OFF_JOB_MANAGERS.

Either call them REGULAR_ONE_OFF_JOB_MANAGERS and PROD_VALIDATION_ONE_OFF_JOB_MANAGERS, or call them ONE_OFF_JOB_MANAGERS and AUDIT_JOB_MANAGERS, or something that suggests more clearly that both categories are disjoint. The "audit" naming might be simpler/shorter, but if you go with that, change the new variable names accordingly.

ankita240796 replied:
Done. I just have a doubt, by the new variable names you mean the variable names introduced for separating the jobs into a new panel or the names of all the jobs and files?

seanlip replied:
More like the former --- so, what you have is fine. The idea is to disambiguate clearly when the two types of jobs are sitting side-by-side.

ankita240796 replied:
Got it, Thanks!

ankita240796 committed:
Merge branch 'develop' into usermodels

seanlip reviewed (approved):
Please fix naming, then this is good to merge. Thanks!

seanlip commented:
@kevinlee12 @nithusha21 @apb7 PTAL as codeowners

@bansalnitish PTAL as mentor

ankita240796 committed:
Rename job

ankita240796 committed:
Remove extra braces

bansalnitish reviewed (approved):
LGTM, Thanks @ankita240796!

apb7 reviewed (approved):
LGTM from code owner's perspective, thanks!

ankita240796 committed:
Add missing test, fix spellings

ankita240796 commented:
Hi @kevinlee12 @nithusha21, can you PTAL. Thanks!

kevinlee12 reviewed (approved):
lgtm! (from code owner's perspective)

seanlip reviewed (approved):
LGTM. @nithusha21 PTAL. Thanks!

ankita240796 commented:
Hi @seanlip @bansalnitish, currently coverage report does not show a 100% coverage for prod jobs files. The lines missed are: 4759, 4693, 4492, 4221, 4163, 1917, 1428, 1368, 1341

The test for each line already exists and it should be shown as covered instead. The lines are of the format:
```
if model is None or model.deleted:
    continue
```

I tested each of the line by adding a raise Exception instead of continue and running the test only for the model to which the line belongs. The exception was raised for all these lines. 

I discussed this with @lilithxxx  as well and he mentioned that he faced a similar issue with continue statement, so one of the fix is to change 

```
if condition:
   continue
code
```

to

```
if not condition:
    code
```

So, is it fine to make these changes for these 9 lines?

Thanks!

seanlip commented:
(I'm going to merge this, so as not to run through the Travis cycle again. If further changes are needed to tests, let's do them in a follow-up PR.)

@ankita240796 -- a quick search uncovered https://bitbucket.org/ned/coveragepy/issues/198/continue-marked-as-not-covered . Can you investigate whether this ever got fixed, or whether it still remains as a bug in the coverage package? Is there a more up-to-date version of that thread?

ankita240796 commented:
> @ankita240796 -- a quick search uncovered https://bitbucket.org/ned/coveragepy/issues/198/continue-marked-as-not-covered . Can you investigate whether this ever got fixed, or whether it still remains as a bug in the coverage package? Is there a more up-to-date version of that thread?

Sure, on it!
